server/src/main/java/com/tencent/rss/server/ShuffleServerGrpcService.java

```diff
@@ -466,11 +476,78 @@ public void getShuffleIndex(GetShuffleIndexRequest request,
       status = StatusCode.INTERNAL_ERROR;
       msg = "Can't require memory to get shuffle index";
       LOG.error(msg + " for " + requestInfo);
-      reply = GetShuffleIndexResponse.newBuilder()
+      reply = GetLocalShuffleIndexResponse.newBuilder()
+          .setStatus(valueOf(status))
+          .setRetMsg(msg)
+          .build();
+    }
+    responseObserver.onNext(reply);
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void getMemoryShuffleData(GetMemoryShuffleDataRequest request,
+      StreamObserver<GetMemoryShuffleDataResponse> responseObserver) {
+    String appId = request.getAppId();
+    int shuffleId = request.getShuffleId();
+    int partitionId = request.getPartitionId();
+    long blockId = request.getLastBlockId();
+    int readBufferSize = request.getReadBufferSize();
+    long start = System.currentTimeMillis();
+    StatusCode status = StatusCode.SUCCESS;
+    String msg = "OK";
+    GetMemoryShuffleDataResponse reply;
+    String requestInfo = "appId[" + appId + "], shuffleId[" + shuffleId + "], partitionId["
+        + partitionId + "]";
+
+    // todo: if can get the exact memory size?
+    if (shuffleServer.getShuffleBufferManager().requireReadMemoryWithRetry(readBufferSize)) {
```

Comment:
why we use read memory? i think it occupy write memory.

Reply:
when get data from shuffle server, always request memory first to avoid OOM, eg,
1. Request 10M data from memory
2. Get Data and it is flushed before send to client
3. Data is send to client
Between step 2&3, there has no any metrics to track 10M, it should be kind of memory leak.